catanatron_core/catanatron/game.py

```diff
@@ -7,16 +7,45 @@
 import sys
 from typing import Iterable, Union
 
-from catanatron.models.enums import Action
+from catanatron.models.enums import Action, ActionPrompt, ActionType
 from catanatron.state import State, apply_action
-from catanatron.state_functions import player_key
+from catanatron.state_functions import player_key, player_has_rolled
 from catanatron.models.map import CatanMap
 from catanatron.models.player import Color, Player
 
 # To timeout RandomRobots from getting stuck...
 TURNS_LIMIT = 1000
 
 
+def is_valid_action(game, action):
+    """True if its a valid action right now. An action is valid
+    if its in playable_actions or if its a OFFER_TRADE in the right time."""
+    if action.action_type == ActionType.OFFER_TRADE:
+        return (
+            game.state.current_color() == action.color
+            and game.state.current_prompt == ActionPrompt.PLAY_TURN
+            and player_has_rolled(game.state, action.color)
+            and is_valid_trade(action.value)
+        )
+
+    return action in game.state.playable_actions
```

Comment:
shouldn't this be taking in the state instead of the game? all the comparisons here are against the game state. (law of demeter)

for a more functional prog. approach, you could consider having function validators which you can pass the game state and action to, and will give you a boolean. Then this function can be something like: `all(validate(state, action) for validate in action_validators)` (where action_validators is an iterable of functions returning a boolean). this can help encapsulate each of the different action_type validators rather than having a mega-method to handle all of the if-clauses.

Reply:
Agreed! Changed to ask for just state.